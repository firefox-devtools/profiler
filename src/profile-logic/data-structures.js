/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import {
  GECKO_PROFILE_VERSION,
  PROCESSED_PROFILE_VERSION,
} from '../app-logic/constants';

import type {
  RawThread,
  RawSamplesTable,
  SamplesTable,
  FrameTable,
  RawStackTable,
  StackTable,
  FuncTable,
  RawMarkerTable,
  JsAllocationsTable,
  UnbalancedNativeAllocationsTable,
  BalancedNativeAllocationsTable,
  ResourceTable,
  NativeSymbolTable,
  Profile,
  ExtensionTable,
  CategoryList,
  JsTracerTable,
  CallNodeTable,
} from 'firefox-profiler/types';

/**
 * This module collects all of the creation of new empty profile data structures.
 */

export function getEmptyStackTable(): StackTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    frame: [],
    prefix: [],
    category: [],
    subcategory: [],
    length: 0,
  };
}

export function getEmptySamplesTable(): RawSamplesTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    weightType: 'samples',
    weight: null,
    stack: [],
    time: [],
    length: 0,
  };
}

export function getEmptyRawStackTable(): RawStackTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    frame: [],
    prefix: [],
    length: 0,
  };
}

/**
 * Returns an empty samples table with eventDelay field instead of responsiveness.
 * eventDelay is a new field and it replaced responsiveness. We should still
 * account for older profiles and use both of the flavors if needed.
 */
export function getEmptySamplesTableWithEventDelay(): RawSamplesTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    weightType: 'samples',
    weight: null,
    eventDelay: [],
    stack: [],
    time: [],
    length: 0,
  };
}

/**
 * Returns an empty samples table with responsiveness field instead of eventDelay.
 * responsiveness is the older field and replaced with eventDelay. We should
 * account for older profiles and use both of the flavors if needed.
 */
export function getEmptySamplesTableWithResponsiveness(): SamplesTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    weightType: 'samples',
    weight: null,
    responsiveness: [],
    stack: [],
    time: [],
    length: 0,
  };
}

export function getEmptyFrameTable(): FrameTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    address: [],
    inlineDepth: [],
    category: [],
    subcategory: [],
    func: [],
    nativeSymbol: [],
    innerWindowID: [],
    implementation: [],
    line: [],
    column: [],
    length: 0,
  };
}

export function shallowCloneFrameTable(frameTable: FrameTable): FrameTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    address: frameTable.address.slice(),
    inlineDepth: frameTable.inlineDepth.slice(),
    category: frameTable.category.slice(),
    subcategory: frameTable.subcategory.slice(),
    func: frameTable.func.slice(),
    nativeSymbol: frameTable.nativeSymbol.slice(),
    innerWindowID: frameTable.innerWindowID.slice(),
    implementation: frameTable.implementation.slice(),
    line: frameTable.line.slice(),
    column: frameTable.column.slice(),
    length: frameTable.length,
  };
}

export function getEmptyFuncTable(): FuncTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    isJS: [],
    relevantForJS: [],
    name: [],
    resource: [],
    fileName: [],
    lineNumber: [],
    columnNumber: [],
    length: 0,
  };
}

export function shallowCloneFuncTable(funcTable: FuncTable): FuncTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    isJS: funcTable.isJS.slice(),
    relevantForJS: funcTable.relevantForJS.slice(),
    name: funcTable.name.slice(),
    resource: funcTable.resource.slice(),
    fileName: funcTable.fileName.slice(),
    lineNumber: funcTable.lineNumber.slice(),
    columnNumber: funcTable.columnNumber.slice(),
    length: funcTable.length,
  };
}

export function shallowCloneNativeSymbolTable(
  nativeSymbols: NativeSymbolTable
): NativeSymbolTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    libIndex: nativeSymbols.libIndex.slice(),
    address: nativeSymbols.address.slice(),
    name: nativeSymbols.name.slice(),
    functionSize: nativeSymbols.functionSize.slice(),
    length: nativeSymbols.length,
  };
}

export function getEmptyResourceTable(): ResourceTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    lib: [],
    name: [],
    host: [],
    type: [],
    length: 0,
  };
}

export function getEmptyNativeSymbolTable(): NativeSymbolTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    libIndex: [],
    address: [],
    name: [],
    functionSize: [],
    length: 0,
  };
}

export function getEmptyRawMarkerTable(): RawMarkerTable {
  // Important!
  // If modifying this structure, please update all callers of this function to ensure
  // that they are pushing on correctly to the data structure. These pushes may not
  // be caught by the type system.
  return {
    data: [],
    name: [],
    startTime: [],
    endTime: [],
    phase: [],
    category: [],
    length: 0,
  };
}

export function getEmptyJsAllocationsTable(): JsAllocationsTable {
  // Important!
  // If modifying this structure, please update all callers of this function to ensure
  // that they are pushing on correctly to the data structure. These pushes may not
  // be caught by the type system.
  return {
    time: [],
    className: [],
    typeName: [],
    coarseType: [],
    weight: [],
    weightType: 'bytes',
    inNursery: [],
    stack: [],
    length: 0,
  };
}

/**
 * The native allocation tables come in two varieties. Get one of the members of the
 * union.
 */
export function getEmptyUnbalancedNativeAllocationsTable(): UnbalancedNativeAllocationsTable {
  // Important!
  // If modifying this structure, please update all callers of this function to ensure
  // that they are pushing on correctly to the data structure. These pushes may not
  // be caught by the type system.
  return {
    time: [],
    weight: [],
    weightType: 'bytes',
    stack: [],
    length: 0,
  };
}

/**
 * The native allocation tables come in two varieties. Get one of the members of the
 * union.
 */
export function getEmptyBalancedNativeAllocationsTable(): BalancedNativeAllocationsTable {
  // Important!
  // If modifying this structure, please update all callers of this function to ensure
  // that they are pushing on correctly to the data structure. These pushes may not
  // be caught by the type system.
  return {
    time: [],
    weight: [],
    weightType: 'bytes',
    stack: [],
    memoryAddress: [],
    threadId: [],
    length: 0,
  };
}

export function shallowCloneRawMarkerTable(
  markerTable: RawMarkerTable
): RawMarkerTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    data: markerTable.data.slice(),
    name: markerTable.name.slice(),
    startTime: markerTable.startTime.slice(),
    endTime: markerTable.endTime.slice(),
    phase: markerTable.phase.slice(),
    category: markerTable.category.slice(),
    length: markerTable.length,
  };
}

export function getResourceTypes() {
  return {
    unknown: 0,
    library: 1,
    addon: 2,
    webhost: 3,
    otherhost: 4,
    url: 5,
  };
}

/**
 * Export a read-only copy of the resource types.
 */
export const resourceTypes = (getResourceTypes(): $Exact<
  $ReadOnly<$Call<typeof getResourceTypes>>,
>);

export function getEmptyExtensions(): ExtensionTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    id: [],
    name: [],
    baseURL: [],
    length: 0,
  };
}

export function getDefaultCategories(): CategoryList {
  return [
    // Make sure 'Other' is at index 0, as it's used as the category for stacks when no
    // categories are provided by an imported (non-Gecko profiler) profile.
    { name: 'Other', color: 'grey', subcategories: ['Other'] },
    { name: 'Idle', color: 'transparent', subcategories: ['Other'] },
    { name: 'Layout', color: 'purple', subcategories: ['Other'] },
    { name: 'JavaScript', color: 'yellow', subcategories: ['Other'] },
    { name: 'GC / CC', color: 'orange', subcategories: ['Other'] },
    { name: 'Network', color: 'lightblue', subcategories: ['Other'] },
    { name: 'Graphics', color: 'green', subcategories: ['Other'] },
    { name: 'DOM', color: 'blue', subcategories: ['Other'] },
  ];
}

export function getEmptyJsTracerTable(): JsTracerTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    events: [],
    timestamps: [],
    durations: [],
    line: [],
    column: [],
    length: 0,
  };
}

export function getEmptyThread(overrides?: $Shape<RawThread>): RawThread {
  const defaultThread: RawThread = {
    processType: 'default',
    processStartupTime: 0,
    processShutdownTime: null,
    registerTime: 0,
    unregisterTime: null,
    pausedRanges: [],
    name: 'Empty',
    isMainThread: false,
    pid: '0',
    tid: 0,
    // Creating samples with event delay since it's the new samples table.
    samples: getEmptySamplesTableWithEventDelay(),
    markers: getEmptyRawMarkerTable(),
    stackTable: getEmptyRawStackTable(),
    frameTable: getEmptyFrameTable(),
    stringArray: [],
    funcTable: getEmptyFuncTable(),
    resourceTable: getEmptyResourceTable(),
    nativeSymbols: getEmptyNativeSymbolTable(),
  };

  return {
    ...defaultThread,
    ...overrides,
  };
}

export function getEmptyProfile(): Profile {
  return {
    meta: {
      interval: 1,
      startTime: 0,
      abi: '',
      misc: '',
      oscpu: '',
      platform: '',
      processType: 0,
      extensions: getEmptyExtensions(),
      categories: getDefaultCategories(),
      product: 'Firefox',
      stackwalk: 0,
      toolkit: '',
      version: GECKO_PROFILE_VERSION,
      preprocessedProfileVersion: PROCESSED_PROFILE_VERSION,
      appBuildID: '',
      sourceURL: '',
      physicalCPUs: 0,
      logicalCPUs: 0,
      CPUName: '',
      symbolicated: true,
      markerSchema: [],
    },
    libs: [],
    pages: [],
    threads: [],
  };
}

export function getEmptyCallNodeTable(): CallNodeTable {
  return {
    // Important!
    // If modifying this structure, please update all callers of this function to ensure
    // that they are pushing on correctly to the data structure. These pushes may not
    // be caught by the type system.
    prefix: new Int32Array(0),
    subtreeRangeEnd: new Uint32Array(0),
    nextSibling: new Int32Array(0),
    func: new Int32Array(0),
    category: new Int32Array(0),
    subcategory: new Int32Array(0),
    innerWindowID: new Float64Array(0),
    sourceFramesInlinedIntoSymbol: new Int32Array(0),
    depth: [],
    maxDepth: -1,
    length: 0,
  };
}
