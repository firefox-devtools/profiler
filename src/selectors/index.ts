/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';
export * from './code';

import * as app from './app';
import {
  selectedThreadSelectors as selectedThread,
  selectedNodeSelectors as selectedNode,
} from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';
import * as code from './code';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  selectedNode,
  l10n,
  cpu,
  code,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
