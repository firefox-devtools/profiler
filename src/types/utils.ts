/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
