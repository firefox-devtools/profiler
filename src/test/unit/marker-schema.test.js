/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
import {
  formatFromMarkerSchema,
  formatMarkupFromMarkerSchema,
  parseLabel,
  markerSchemaFrontEndOnly,
} from '../../profile-logic/marker-schema';
import type { MarkerSchema, Marker } from 'firefox-profiler/types';
import { getDefaultCategories } from '../../profile-logic/data-structures';
import { storeWithProfile } from '../fixtures/stores';
import { getMarkerSchema } from '../../selectors/profile';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';
import { markerSchemaForTests } from '../fixtures/profiles/marker-schema';
import { StringTable } from '../../utils/string-table';

/**
 * Generally, higher level type of testing is preferred to detailed unit tests of
 * implementation behavior, but the marker schema labels use a custom mini-parser, and
 * it would be easy for them to have errors. These tests cover a variety of different
 * code branches, especially parse errors.
 */
describe('marker schema labels', function () {
  type LabelOptions = {|
    schemaData: $PropertyType<MarkerSchema, 'data'>,
    label: string,
    payload: any,
  |};

  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  function applyLabel(options: LabelOptions): string {
    const { schemaData, label, payload } = options;
    const categories = getDefaultCategories();
    const stringTable = StringTable.withBackingArray([
      'IPC Message',
      'MouseDown Event',
    ]);

    const schema = {
      name: 'TestDefinedMarker',
      display: [],
      data: schemaData,
    };

    const marker: Marker = {
      start: 2,
      end: 5,
      name: 'TestDefinedMarker',
      category: 0,
      threadId: 1,
      data: payload,
    };
    const getter = parseLabel(schema, categories, stringTable, label);

    // There is only one marker, marker 0
    return getter(marker);
  }

  it('can parse very simple labels', function () {
    expect(
      applyLabel({
        label: 'Just text',
        schemaData: [],
        payload: {},
      })
    ).toEqual('Just text');
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  it('can parse a label with just a lookup value', function () {
    expect(
      applyLabel({
        label: '{marker.data.duration}',
        schemaData: [{ key: 'duration', label: 'Duration', format: 'seconds' }],
        payload: { duration: 12345 },
      })
    ).toEqual('12.345s');
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  it('can parse a label with surrounding text', function () {
    expect(
      applyLabel({
        label: 'It took {marker.data.duration} for this test.',
        schemaData: [{ key: 'duration', label: 'Duration', format: 'seconds' }],
        payload: { duration: 12345 },
      })
    ).toEqual('It took 12.345s for this test.');
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  it('can mix and match lookups', function () {
    expect(
      applyLabel({
        label: 'It took {marker.data.duration}, which is {marker.data.ratio}',
        schemaData: [
          { key: 'duration', label: 'Duration', format: 'seconds' },
          { key: 'ratio', label: 'Ratio', format: 'percentage' },
        ],
        payload: {
          duration: 12345,
          ratio: 0.12345,
        },
      })
    ).toEqual('It took 12.345s, which is 12%');
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  it('is empty if there is no information in a payload', function () {
    expect(
      applyLabel({
        label: 'This will be nothing: "{marker.data.nokey}"',
        schemaData: [{ key: 'duration', label: 'Duration', format: 'seconds' }],
        payload: { duration: 12345 },
      })
    ).toEqual('This will be nothing: ""');
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  it('can look up various parts of the marker', function () {
    const text = applyLabel({
      label: [
        'Start: {marker.start}',
        'End: {marker.end}',
        'Duration: {marker.duration}',
        'Name: {marker.name}',
        'Category: {marker.category}',
      ].join('\n'),
      schemaData: [],
      payload: {},
    });

    expect(text.split('\n')).toEqual([
      'Start: 2ms',
      'End: 5ms',
      'Duration: 3ms',
      'Name: TestDefinedMarker',
      'Category: Other',
    ]);
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  it('can parse labels with unique strings', function () {
    expect(
      applyLabel({
        label: '{marker.data.message} happened because of {marker.data.event}',
        schemaData: [
          { key: 'message', label: 'Message', format: 'unique-string' },
          { key: 'event', label: 'Event', format: 'unique-string' },
        ],
        payload: {
          message: 0,
          event: 1,
        },
      })
    ).toEqual('IPC Message happened because of MouseDown Event');
    expect(console.error).toHaveBeenCalledTimes(0);
  });

  describe('parseErrors', function () {
    function testParseError(label: string) {
      expect(
        applyLabel({
          label,
          schemaData: [
            { key: 'duration', label: 'Duration', format: 'seconds' },
          ],
          payload: { duration: 12345 },
        })
      ).toEqual('Parse error: ""');
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error.mock.calls).toMatchSnapshot();
    }

    // eslint-disable-next-line jest/expect-expect
    it('errors if not looking up into a marker', function () {
      testParseError('Parse error: "{duration}"');
    });

    // eslint-disable-next-line jest/expect-expect
    it('errors if looking up into a part of the marker that does not exist', function () {
      testParseError('Parse error: "{marker.nothing}"');
    });

    // eslint-disable-next-line jest/expect-expect
    it('errors when accessing random properties', function () {
      testParseError('Parse error: "{property.value}"');
    });

    // eslint-disable-next-line jest/expect-expect
    it('errors when accessing twice into a payload', function () {
      testParseError('Parse error: "{marker.data.duration.extra}"');
    });
  });
});

describe('marker schema formatting', function () {
  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});
  });

  it('can apply a variety of formats', function () {
    const entries = [
      ['url', 'http://example.com'],
      ['file-path', '/Users/me/gecko'],
      ['file-path', null],
      ['file-path', undefined],
      ['sanitized-string', 'domain.name'],
      ['duration', 0],
      ['duration', 10],
      ['duration', 12.3456789],
      ['duration', 123456.789],
      ['duration', 23456789],
      ['duration', 50000],
      ['duration', 2000000],
      ['duration', 50000000],
      ['duration', 123456789],
      ['duration', 0.000123456],
      ['duration', 1000 * 60 - 501], // slightly less than 1min, should not be shown as '60s'
      ['duration', 1000 * 60 - 500], // slightly less than 1min, should not be shown as '60s'
      ['duration', 1000 * 60 + 1], // slightly more than 1min, should not be shown as '1min0s'
      ['duration', 779873.639], // '13min60s' should be rounded to '14min'
      ['duration', 1000 * 3600 - 501], // 1h - 501ms slightly less than 1h, should be shown as '59min59s'
      ['duration', 1000 * 3600 - 500], // 1h - 500s slightly less than 1h, should be shown as '1h', not '0h60min'
      ['duration', 1000 * 3600 + 1], // slightly more than 1h, should not be shown as '1h0min'
      ['duration', 1000 * 3600 * 24 - 31 * 1000], // 1d - 31s, should be shown as '23h59min'
      ['duration', 1000 * 3600 * 24 - 30 * 1000], // 1d - 30s, should be shown as '1d', not '0d24h'
      ['duration', 1000 * 3600 * 24 + 1], // slightly more than 1 day, should not be shown as '1d0h'
      ['time', 12.3456789],
      ['seconds', 0],
      ['seconds', 10],
      ['seconds', 12.3456789],
      ['seconds', 123456.789],
      ['seconds', 0.000123456],
      ['milliseconds', 0],
      ['milliseconds', 10],
      ['milliseconds', 12.3456789],
      ['milliseconds', 123456.789],
      ['milliseconds', 0.000123456],
      ['microseconds', 0],
      ['microseconds', 10],
      ['microseconds', 12.3456789],
      ['microseconds', 123456.789],
      ['microseconds', 0.000123456],
      ['nanoseconds', 0],
      ['nanoseconds', 10],
      ['nanoseconds', 12.3456789],
      ['nanoseconds', 123456.789],
      ['nanoseconds', 0.000123456],
      ['bytes', 0],
      ['bytes', 10],
      ['bytes', 12.3456789],
      ['bytes', 123456.789],
      ['bytes', 123456789],
      ['bytes', 123456789123],
      ['bytes', 0.000123456],
      ['integer', 0],
      ['integer', 10],
      ['integer', 12.3456789],
      ['integer', 123456.789],
      ['integer', 0.000123456],
      ['decimal', 0],
      ['decimal', 10],
      ['decimal', 12.3456789],
      ['decimal', 123456.789],
      ['decimal', 0.000123456],
      ['percentage', 0],
      ['percentage', 0.1],
      ['percentage', 0.123456789],
      ['percentage', 1234.56789],
      ['percentage', 0.000123456],
      ['pid', '2322'],
      ['tid', '2427'],
      ['unique-string', 0], // Should be "IPC Message", see "stringTable" in "applyLabel" fn
      ['unique-string', 1], // Should be "MouseDown Event", see "stringTable" in "applyLabel" fn
      ['unique-string', null], // Should be "(empty)"
      ['unique-string', undefined], // Should be "(empty)"
      ['unique-string', 42], // Should be "(empty)"
    ];

    expect(
      entries.map(
        ([format, value]) =>
          format +
          ' - ' +
          formatFromMarkerSchema(
            'none',
            format,
            value,
            StringTable.withBackingArray(['IPC Message', 'MouseDown Event'])
          )
      )
    ).toMatchInlineSnapshot(`
      Array [
        "url - http://example.com",
        "file-path - /Users/me/gecko",
        "file-path - (empty)",
        "file-path - (empty)",
        "sanitized-string - domain.name",
        "duration - 0s",
        "duration - 10ms",
        "duration - 12.346ms",
        "duration - 2m3s",
        "duration - 6h31m",
        "duration - 50s",
        "duration - 33m20s",
        "duration - 13h53m",
        "duration - 1d10h",
        "duration - 123.46ns",
        "duration - 59.499s",
        "duration - 1m",
        "duration - 1m",
        "duration - 13m",
        "duration - 59m59s",
        "duration - 1h",
        "duration - 1h",
        "duration - 23h59m",
        "duration - 1d",
        "duration - 1d",
        "time - 12.346ms",
        "seconds - 0s",
        "seconds - 0.010s",
        "seconds - 0.012s",
        "seconds - 123.46s",
        "seconds - 0.000s",
        "milliseconds - 0ms",
        "milliseconds - 10ms",
        "milliseconds - 12ms",
        "milliseconds - 123,457ms",
        "milliseconds - 0.000ms",
        "microseconds - 0μs",
        "microseconds - 10μs",
        "microseconds - 12μs",
        "microseconds - 123,457μs",
        "microseconds - 0.000μs",
        "nanoseconds - 0ns",
        "nanoseconds - 10.0ns",
        "nanoseconds - 12.3ns",
        "nanoseconds - 123,457ns",
        "nanoseconds - 0.0001ns",
        "bytes - 0B",
        "bytes - 10B",
        "bytes - 12B",
        "bytes - 121KB",
        "bytes - 118MB",
        "bytes - 115GB",
        "bytes - 0.000B",
        "integer - 0",
        "integer - 10",
        "integer - 12",
        "integer - 123,457",
        "integer - 0",
        "decimal - 0",
        "decimal - 10",
        "decimal - 12",
        "decimal - 123,457",
        "decimal - 0.000",
        "percentage - 0%",
        "percentage - 10%",
        "percentage - 12%",
        "percentage - 123,457%",
        "percentage - 0.0%",
        "pid - PID: 2322",
        "tid - TID: 2427",
        "unique-string - IPC Message",
        "unique-string - MouseDown Event",
        "unique-string - (empty)",
        "unique-string - (empty)",
        "unique-string - (empty)",
      ]
    `);
  });

  it('supports complex formats', function () {
    const entries = [
      ['url', 'http://example.com'],
      ['file-path', '/Users/me/gecko'],
      ['file-path', null],
      ['file-path', undefined],
      ['duration', 0],
      ['duration', 10],
      ['duration', 12.3456789],
      [
        { type: 'table', columns: [{ type: 'string' }, { type: 'integer' }] },
        [
          ['a', 1],
          ['b', 2],
        ],
      ],
      [
        {
          type: 'table',
          columns: [
            { type: 'string', label: 'a' },
            { type: 'integer', label: 'b' },
          ],
        },
        [['b', 2]],
      ],
      [
        {
          type: 'table',
          columns: [{ type: 'string', label: 'a' }, { type: 'integer' }],
        },
        [['b', 2]],
      ],
      [
        { type: 'table', columns: [{ type: 'string', label: 'a' }, {}] },
        [['b', 2]],
      ],
      ['list', []],
      ['list', ['a', 'b']],
    ];
    const stringTable = StringTable.withBackingArray([]);
    expect(
      entries.map(([format, value]) => [
        format,
        value,
        formatMarkupFromMarkerSchema('none', format, value, stringTable),
        formatFromMarkerSchema('none', format, value, stringTable),
      ])
    ).toMatchSnapshot();
  });
});

describe('getMarkerSchema', function () {
  it('combines front-end and Gecko marker schema', function () {
    const { profile } = getProfileFromTextSamples('A');
    profile.meta.markerSchema = markerSchemaForTests;
    const { getState } = storeWithProfile(profile);
    const combinedSchema = getMarkerSchema(getState());

    // Find front-end only marker schema.
    expect(
      profile.meta.markerSchema.find((schema) => schema.name === 'Jank')
    ).toBeUndefined();
    expect(
      markerSchemaFrontEndOnly.find((schema) => schema.name === 'Jank')
    ).toBeTruthy();
    expect(
      combinedSchema.find((schema) => schema.name === 'Jank')
    ).toBeTruthy();

    // Find the Gecko only marker schema.
    expect(
      profile.meta.markerSchema.find((schema) => schema.name === 'GCMajor')
    ).toBeTruthy();
    expect(
      markerSchemaFrontEndOnly.find((schema) => schema.name === 'GCMajor')
    ).toBeUndefined();
    expect(
      combinedSchema.find((schema) => schema.name === 'GCMajor')
    ).toBeTruthy();
  });
});
