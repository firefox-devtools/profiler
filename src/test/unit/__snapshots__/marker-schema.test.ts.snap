// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`marker schema formatting supports complex formats 1`] = `
Array [
  Array [
    "url",
    "http://example.com",
    <a
      className="marker-link-value"
      href="http://example.com"
      rel="noreferrer"
      target="_blank"
    >
      example.com
    </a>,
    "http://example.com",
  ],
  Array [
    "file-path",
    "/Users/me/gecko",
    "/Users/me/gecko",
    "/Users/me/gecko",
  ],
  Array [
    "file-path",
    null,
    "(empty)",
    "(empty)",
  ],
  Array [
    "file-path",
    undefined,
    "(empty)",
    "(empty)",
  ],
  Array [
    "duration",
    0,
    "0s",
    "0s",
  ],
  Array [
    "duration",
    10,
    "10ms",
    "10ms",
  ],
  Array [
    "duration",
    12.3456789,
    "12.346ms",
    "12.346ms",
  ],
  Array [
    Object {
      "columns": Array [
        Object {
          "type": "string",
        },
        Object {
          "type": "integer",
        },
      ],
      "type": "table",
    },
    Array [
      Array [
        "a",
        1,
      ],
      Array [
        "b",
        2,
      ],
    ],
    <table
      className="marker-table-value"
    >
      <tbody>
        <tr>
          <td>
            a
          </td>
          <td>
            1
          </td>
        </tr>
        <tr>
          <td>
            b
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>,
    "(a, 1),(b, 2)",
  ],
  Array [
    Object {
      "columns": Array [
        Object {
          "label": "a",
          "type": "string",
        },
        Object {
          "label": "b",
          "type": "integer",
        },
      ],
      "type": "table",
    },
    Array [
      Array [
        "b",
        2,
      ],
    ],
    <table
      className="marker-table-value"
    >
      <thead>
        <tr>
          <th>
            a
          </th>
          <th>
            b
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            b
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>,
    "(a, b),(b, 2)",
  ],
  Array [
    Object {
      "columns": Array [
        Object {
          "label": "a",
          "type": "string",
        },
        Object {
          "type": "integer",
        },
      ],
      "type": "table",
    },
    Array [
      Array [
        "b",
        2,
      ],
    ],
    <table
      className="marker-table-value"
    >
      <thead>
        <tr>
          <th>
            a
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            b
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>,
    "(a, (empty)),(b, 2)",
  ],
  Array [
    Object {
      "columns": Array [
        Object {
          "label": "a",
          "type": "string",
        },
        Object {},
      ],
      "type": "table",
    },
    Array [
      Array [
        "b",
        2,
      ],
    ],
    <table
      className="marker-table-value"
    >
      <thead>
        <tr>
          <th>
            a
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            b
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>,
    "(a, (empty)),(b, 2)",
  ],
  Array [
    "list",
    Array [],
    <ul
      className="marker-list-value"
    />,
    "",
  ],
  Array [
    "list",
    Array [
      "a",
      "b",
    ],
    <ul
      className="marker-list-value"
    >
      <li>
        a
      </li>
      <li>
        b
      </li>
    </ul>,
    "a, b",
  ],
]
`;

exports[`marker schema labels parseErrors errors if looking up into a part of the marker that does not exist 1`] = `
Array [
  Array [
    "Error processing the label \\"Parse error: \\"{marker.nothing}\\"\\" because of the marker.nothing. Currently the labels in the marker schema take the form of \\"marker.data.keyName\\" or \\"marker.startTime\\". No other type of access is currently supported.",
  ],
]
`;

exports[`marker schema labels parseErrors errors if not looking up into a marker 1`] = `
Array [
  Array [
    "Error processing the label \\"Parse error: \\"{duration}\\"\\" because of the duration. Currently the labels in the marker schema take the form of \\"marker.data.keyName\\" or \\"marker.startTime\\". No other type of access is currently supported.",
  ],
]
`;

exports[`marker schema labels parseErrors errors when accessing random properties 1`] = `
Array [
  Array [
    "Error processing the label \\"Parse error: \\"{property.value}\\"\\" because of the property.value. Currently the labels in the marker schema take the form of \\"marker.data.keyName\\" or \\"marker.startTime\\". No other type of access is currently supported.",
  ],
]
`;

exports[`marker schema labels parseErrors errors when accessing twice into a payload 1`] = `
Array [
  Array [
    "Error processing the label \\"Parse error: \\"{marker.data.duration.extra}\\"\\" because of the marker.data.duration.extra. Currently the labels in the marker schema take the form of \\"marker.data.keyName\\" or \\"marker.startTime\\". No other type of access is currently supported.",
  ],
]
`;
